### DISCUSSION TITLE
Using SHOP variables in Lisp

### DISCUSSION THREAD

**Discussion by ko56** [shop-planner/shop3#126] [Q&A] [answered]
Is it possible to access the value of a SHOP variable, say that of an atom, in a Lisp function?

**Comment by rpgoldman** [comment]
You can do this in the context of the theorem-prover.  In the theorem-prover, what actually happens is the value of a SHOP variable can be *substituted* into an expression that is interpreted as lisp code.

For example from a board game playing program:

```
(:- (distance ?location1 ?location2 ?distance)
    ((hex-location ?location1 ?hex1)
     (hex-location ?location2 ?hex2)
     (assign ?distance (hex-distance ?hex1 ?hex2)))
```

Here we have used collected the coordinate values `?hex1` and `?hex2` and then they are passed into a lisp function, `hex-distance`. 

Note an obscure issue here -- if these coordinates were, for some reason, lisp symbols, like `'A12` and `'B14`(instead of, say `0112`  and `0214`) then the above would cause an error, because the values would be plugged in, and then `hex-distance` would be invoked, and `A12` and `B14` would look like unbound variables to Lisp.

In cases like that, you would have to modify the above to use:
```
     (assign ?distance (hex-distance (quote ?hex1) (quote ?hex2))))
```

**Reply by ko56** [reply]
Thanks, I'm already doing things like the example you gave.  My question was not properly asked.  
What I meant to ask is if there's a way to access the atoms without passing them as arguments to the Lisp function.
E.g. I have a predicate "greater", and I have defined/set
(greater 2 1)
(greater b a)
...
These atoms are part of the current state, so maybe that way, somehow?

**Reply by rpgoldman** [reply, marked as answer]
I'm still not sure I understand what you mean by a SHOP variable.  SHOP variables are internal to the theorem-prover, and so are only accessible through mechanisms like the one I described earlier.

If you want to access _atoms_ from Lisp code, you can use the special `*current-state*` variable. See section 3.4 in the manual.

**Reply by ko56** [reply]
I was using the term "variable" very loosely.  I should have said "atom".
I know about *current-state*, but I never paid attention to its structure.
I will do so now.
